Mathlib/GroupTheory/GroupAction/Jordan.lean

```diff
@@ -0,0 +1,475 @@
+        ofFixingSubgroup_of_singleton_bijective.surjective hprim.toIsPretransitive
+    · intro t h
+      rw [zero_add, Nat.cast_ofNat, ← one_add_one_eq_two,
+        (ENat.add_left_injective_of_ne_top ENat.one_ne_top).eq_iff] at h
+      obtain ⟨b, htb⟩ := Set.encard_eq_one.mp h
+      obtain ⟨g, hg⟩ := exists_smul_eq G a b
+      have hst : g • ({a} : Set α) = ({b} : Set α) := by
+        rw [Set.smul_set_singleton, hg]
+      rw [htb]
+      refine IsPreprimitive.of_surjective
+        (conjMap_ofFixingSubgroup_bijective (hst := hst)).surjective
+  -- Induction step
+  suffices ∃ (a : α) (t : Set (SubMulAction.ofStabilizer G a)),
+    a ∈ s ∧ s = insert a (Subtype.val '' t) by
+    obtain ⟨a, t, _, hst⟩ := this
+    have ha' : a ∉ Subtype.val '' t := by
+      intro h; rw [Set.mem_image] at h ; obtain ⟨x, hx⟩ := h
+      apply x.prop; rw [hx.right]; exact Set.mem_singleton a
+    have ht_prim : IsPreprimitive (stabilizer G a) (SubMulAction.ofStabilizer G a) := by
+      rw [← is_one_preprimitive_iff]
+      rw [← isMultiplyPreprimitive_succ_iff_ofStabilizer]
+      · apply is_two_preprimitive hG hsn hsn' hprim
+      · norm_num
+    have : IsPreprimitive ↥(fixingSubgroup G (insert a (Subtype.val '' t)))
+        (ofFixingSubgroup G (insert a (Subtype.val '' t))) :=
+      IsPreprimitive.of_surjective
+        (ofFixingSubgroup_of_eq_bijective (hst := hst)).surjective
+    have hGs' : IsPreprimitive (fixingSubgroup (stabilizer G a) t)
+      (ofFixingSubgroup (stabilizer G a) t) :=
+      IsPreprimitive.of_surjective
+        ofFixingSubgroup_insert_map_bijective.surjective
+    rw [isMultiplyPreprimitive_succ_iff_ofStabilizer G (a := a) _ (Nat.le_add_left 1 (n + 1))]
+    refine hrec ht_prim ?_ ?_ hGs' Subtype.finite
+    · -- t.card = Nat.succ n
+      rw [← Set.ncard_image_of_injective t Subtype.val_injective]
+      apply Nat.add_right_cancel
+      rw [← Set.ncard_insert_of_notMem ha', ← hst, hsn]
+    · -- n + 2 < Nat.card (SubMulAction.ofStabilizer G α a)
+      rw [← Nat.add_lt_add_iff_right, nat_card_ofStabilizer_add_one_eq]
+      exact hsn'
+  -- ∃ a t, a ∈ s ∧ s = insert a (Subtype.val '' t)
+  suffices s.Nonempty by
+    obtain ⟨a, ha⟩ := this
+    use a, Subtype.val ⁻¹' s, ha
+    ext x
+    by_cases hx : x = a <;> simp [hx, mem_ofStabilizer_iff, ha]
+  rw [← Set.ncard_pos, hsn]; apply Nat.succ_pos
+
+end Jordan
+
+section Subgroups
+
+namespace Equiv.Perm
+
+open Equiv Set
+
+variable {α : Type*}
+
+variable {G : Subgroup (Perm α)}
+
+theorem subgroup_eq_top_of_nontrivial [Finite α] (hα : Nat.card α ≤ 2) (hG : Nontrivial G) :
+    G = (⊤ : Subgroup (Perm α)) := by
+  apply Subgroup.eq_top_of_le_card
+  rw [Nat.card_perm]
+  apply (Nat.factorial_le hα).trans
+  rwa [Nat.factorial_two, Nat.succ_le, one_lt_card_iff_ne_bot, ← nontrivial_iff_ne_bot]
+
+theorem nontrivial_on_equiv_perm_two {K : Type*} [Group K] [MulAction K α]
+    (hα : Nat.card α = 2) (hK : fixedPoints K α ≠ .univ) :
+    IsMultiplyPretransitive K α 2 := by
```

Comment:
You actually have `IsMultiplyPretransitive K α n` for all `n`, right?

Reply:
Good point!